cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -1828,6 +1828,148 @@ void AddUtf8Length(FunctionRegistry* registry) {
   DCHECK_OK(registry->AddFunction(std::move(func)));
 }
 
+// binary join
+
+template <typename Type>
+struct Join {
```

Comment:
Note that the computation logic doesn't depend on utf8 vs. binary, hence it would be nice to only instantiate two kernel routines (one per offset type), not four.